### PR TITLE
Keyboard service impl

### LINKED ISSUE #315: keyboard service impl
Implement the keyboard grpc service

### PR DESCRIPTION
**Pull request by matthew-hagemann**
Closes #315

### CODE REVIEW

**Comment by matthew-hagemann** [comment]
@d-loose could you let me know if there are any additional endpoints you'd need exposed in order to plug this into `ubuntu_init` (and gets, etc.) :slightly_smiling_face:

**Comment by d-loose** [comment]
I'm afraid we'll need a [`GetKeyboard()`](https://github.com/canonical/ubuntu-desktop-provision/blob/d809a462038d5c4712250d1bfe28972caca03705/packages/ubuntu_bootstrap/lib/services/keyboard_service.dart#L10) method that returns the current keyboard settings and a list of available layouts/variants like [here](https://github.com/canonical/ubuntu-desktop-provision/blob/d809a462038d5c4712250d1bfe28972caca03705/packages/subiquity_client/lib/src/types.dart#L220-L224)

File: provd/protos/keyboard.proto [hunk 1]
```diff
@@ -23,3 +24,23 @@ message SetKeyboardRequest {
...
+    repeated KeyboardLayout layouts = 2;
+}
+
+message KeyboardLayout {
+    string code = 1;
+    string name = 2;
+    repeated KeybaordVariant variants = 3;
+}
+
+message KeybaordVariant {
```
Comment (d-loose) [inline provd/protos/keyboard.proto:43]:
```suggestion
message KeyboardVariant {
```

File: provd/protos/keyboard.proto [hunk 2]
```diff
@@ -23,3 +24,23 @@ message SetKeyboardRequest {
...
+
+message KeyboardSetup {
+    KeyboardSettings settings = 1;
+    repeated KeyboardLayout layouts = 2;
+}
+
+message KeyboardLayout {
+    string code = 1;
+    string name = 2;
+    repeated KeybaordVariant variants = 3;
```
Comment (d-loose) [inline provd/protos/keyboard.proto:40]:
```suggestion
    repeated KeyboardVariant variants = 3;
```

File: provd/internal/services/keyboard/keyboard.go [hunk 3]
```diff
@@ -135,3 +191,49 @@ func (s *Service) SetKeyboard(ctx context.Context, req *pb.SetKeyboardRequest) (
...
+
+	// Create keyboard layout pb structs
+	for code, layout := range rawLayouts {
+		l := &pb.KeyboardLayout{
+			Code: code,
+			Name: layout["layout"],
+		}
+		if variants, ok := layout["variant"]; ok && variants != "" {
+			for _, v := range strings.Split(variants, ",") {
+				l.Variants = append(l.Variants, &pb.KeybaordVariant{
```
Comment (d-loose) [inline provd/internal/services/keyboard/keyboard.go]:
```suggestion
				l.Variants = append(l.Variants, &pb.KeyboardVariant{
```

Comment (d-loose) [reply provd/internal/services/keyboard/keyboard.go]: Sorry, that's already fixed - ran into it in a merge ..

**Comment by d-loose** [comment]
Thanks, great progress!
I think what's still missing are the localized names for the layouts and variants (the list of variants in the keyboard-configuration.yaml seems incomplete). Subiquity has all those json files in `kbds/` and returns a response like this:

```
$ curl -sS -X GET --unix-socket ./packages/subiquity_client/subiquity/.subiquity/socket "http://localhost/keyboard" | jq
{
  "setting": {
    "layout": "us",
    "variant": "",
    "toggle": null
  },
  "layouts": [
...
    {
      "code": "fr",
      "name": "French",
      "variants": [
        {
          "code": "",
          "name": "French"
        },
        {
          "code": "afnor",
          "name": "French - French (AFNOR standardized AZERTY)"
        },
        {
          "code": "azerty",
          "name": "French - French (AZERTY)"
        },
       ...
        {
          "code": "oci",
          "name": "French - Occitan"
        }
      ]
    },
```
 whereas provd currently responds with
```
$ grpcurl -plaintext -unix -proto provd/protos/keyboard.proto -d '{}' /run/provd/provd.sock keyboard.KeyboardService/GetKeyboard
{
  "setup": {
    "settings": {
      "layout": "us",
      "variant": "altgr-intl"
    },
    "layouts": [
 ...
      {
        "code": "fr_FR.UTF-8",
        "name": "fr",
        "variants": [
          {
            "code": "latin9",
            "name": "latin9"
          }
        ]
      },
...
```

image

**Comment by matthew-hagemann** [comment]
@d-loose I mistook which files I needed to parse, and was not swapping based on locale!

I have made the following changes:
- Parsing correct `kbds` file based on locale
- Defaults to `C` (en) if current locale is not in list of available `jsonl` files

I am hoping this now has feature parity with the current installer, let me know if I have missed something :slightly_smiling_face: 

Outputs now look like this:

```
❯ grpcurl -plaintext -unix -proto keyboard.proto -d '{}' /home/matt/test.sock keyboard.KeyboardService/GetKeyboard
{
  "setup": {
    "settings": {
      "layout": "us"
    },
    "layouts": [
      {
        "code": "af",
        "name": "Afghani",
        "variants": [
          {
            "name": "Afghani"
          },
          {
            "code": "ps",
            "name": "Afghani - Pashto"
          },
          {
            "code": "olpc-ps",
            "name": "Afghani - Pashto (Afghanistan, OLPC)"
          },
          {
            "code": "fa-olpc",
            "name": "Afghani - Persian (Afghanistan, Dari OLPC)"
          },
          {
            "code": "uz",
            "name": "Afghani - Uzbek (Afghanistan)"
          },
          {
            "code": "uz-olpc",
            "name": "Afghani - Uzbek (Afghanistan, OLPC)"
          }
        ]
      },

```

I also believe that changing locale and calling `GetKeyboard` again should return a updated list of keyboards if we support the `jsonl`.

**Comment by d-loose** [comment]
Awesome - we're getting there!

But of course there are more issues :D
Under wayland I'm getting the following when trying to set the keyboard from ubuntu-desktop-init:
```
provd[210437]: failed to commit changes to dconf: Cannot autolaunch D-Bus without X11 $DISPLAY
```

**Comment by d-loose** [comment]
Ok, seems to be a systemd socket activation issue. Everything works if I run `provd` from the user session and point ubuntu-desktop-init to the socket :tada:

**Review by d-loose — APPROVED** [review]
Previous comments regarding systemd socket activation still apply - no issues when running it locally and interacting with the socket directly though.
LGTM :+1:

**Pull request merged**